public/app/features/logs/utils.ts

```diff
@@ -343,3 +343,48 @@ export function createLogRowsMap() {
     return false;
   };
 }
+
+function getLabelTypeFromFrame(labelKey: string, frame: DataFrame, index: number): null | string {
+  const typeField = frame.fields.find((field) => field.name === 'labelTypes')?.values[index];
+  if (!typeField) {
+    return null;
+  }
+  return typeField[labelKey] ?? null;
+}
+
+export function getLabelTypeFromRow(label: string, row: LogRowModel) {
+  if (!row.datasourceType) {
+    return null;
+  }
+  const idField = row.dataFrame.fields.find((field) => field.name === 'id');
+  if (!idField) {
+    return null;
+  }
+  const rowIndex = idField.values.findIndex((id) => id === row.rowId);
+  if (rowIndex < 0) {
+    return null;
+  }
+  const labelType = getLabelTypeFromFrame(label, row.dataFrame, rowIndex);
+  if (!labelType) {
+    return null;
+  }
+  return resolveDataSourceLabelType(labelType, row.datasourceType);
+}
+
+function resolveDataSourceLabelType(labelType: string, datasourceType: string) {
```

Comment:
```suggestion
  return getDataSourceLabelType(labelType, row.datasourceType);
}

function getDataSourceLabelType(labelType: string, datasourceType: string) {
```

Just a nit, but `resolve` sounds too much like a `Promise` in my opinion :D